src/contexts/TodosContext.tsx

```diff
@@ -0,0 +1,287 @@
+
+      setTodos(currentTodos => {
+        const index = currentTodos.findIndex(todo => todo.id === id);
+        const updatedTodos = [...currentTodos];
+
+        updatedTodos.splice(index, 1, updatedTodo);
+
+        return updatedTodos;
+      });
+    } catch {
+      setError(UNABLE_UPDATE_ERROR_MESSAGE);
+      throw new Error();
+    } finally {
+      setIsInputFocused(false);
+      setTodosIdsUpdating([]);
+    }
+  };
+
+  const handleToggleTodo = async () => {
+    const todosToChange = todos.some(todo => !todo.completed)
+      ? filterBy(todos, FilterType.Active)
+      : todos;
+    const todosId = todosToChange.map(todo => todo.id);
+
+    setTodosIdsUpdating(todosId);
+
+    const changePromises = todosToChange
+      .map(async (todo) => {
+        try {
+          return await updateTodo(todo.id, { completed: !todo.completed });
+        } catch {
+          return null;
+        }
+      });
+
+    const changedTodos = await Promise.all(changePromises);
+
+    if (changedTodos.includes(null)) {
+      setError(UNABLE_UPDATE_ERROR_MESSAGE);
+    }
+
+    setTodos((prevState) => {
+      return prevState.map((todo) => {
+        const changed = changedTodos
+          .find((changedTodo) => {
+            if (!changedTodo) {
+              return false;
+            }
+
+            return changedTodo.id === todo.id;
+          });
+
+        if (changed) {
+          return changed;
+        }
+
+        return todo;
+      });
+    });
+
+    setIsInputFocused(false);
+    setTodosIdsUpdating([]);
+  };
+
+  const handleDeleteTodo = async (id: number) => {
+    setTodosIdsUpdating([id]);
+
+    try {
+      await deleteTodo(id);
+      setTodos((prevState) => prevState.filter((todo) => todo.id !== id));
+    } catch {
+      setError(UNABLE_DELETE_ERROR_MESSAGE);
+    } finally {
+      setTodosIdsUpdating([]);
+    }
+  };
+
+  const handleDeleteCompletedTodo = async () => {
+    setErrorMessage(null);
+    const completedTodosId = todos
+      .filter(todo => todo.completed)
+      .map(todo => todo.id);
+
+    setTodosIdsUpdating(completedTodosId);
+
+    const deletePromises = completedTodosId.map(async (id) => {
+      try {
+        await deleteTodo(id);
+
+        return id;
+      } catch {
+        return null;
+      }
+    });
+
+    const deletedIds = await Promise.all(deletePromises);
+
+    if (deletedIds.includes(null)) {
+      setError(UNABLE_DELETE_ERROR_MESSAGE);
+    }
+
+    setTodos(
+      prevState => prevState.filter(todo => !deletedIds.includes(todo.id)),
+    );
+
+    setIsInputFocused(true);
+    setTodosIdsUpdating([]);
+  };
+
+  return (
+    <TodosContext.Provider
+      value={{
+        todos,
+        filterType,
+        tempTodo,
+        todosIdsUpdating,
+        errorMessage,
+        visibleTodos,
+        handleFilterChange,
+        handleAddTodo,
+        handleChangeTodo,
+        handleDeleteTodo,
+        handleDeleteCompletedTodo,
+        handleToogleTodo: handleToggleTodo,
```

Comment:
Why?
Use just `handleToggleTodo` right away, because `toogle` it`s not correct.